lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/JsonRpcResponse.kt

```diff
@@ -22,27 +18,40 @@ internal data class JsonRpcResponse<T>(
     val result: T? = null,
 
     @SerialName("error")
-    val error: JsonElement? = null, // FIXME: Add error types
+    val error: JsonRpcError? = null,
 )
 
-internal class JsonRpcResponseDeserializer<T>(private val dataSerializer: KSerializer<T>) : DeserializationStrategy<T> {
-    override val descriptor: SerialDescriptor
-        get() = dataSerializer.descriptor
+@Serializable
+internal data class JsonRpcError(
+    @SerialName("code")
+    val code: Int,
+
+    @SerialName("message")
+    val message: String,
+)
 
-    override fun deserialize(decoder: Decoder): T {
-        require(decoder is JsonDecoder)
+internal fun <T> buildJsonHttpDeserializer(deserializationStrategy: KSerializer<T>): HttpResponseDeserializer<T> {
+    return { response ->
+        if (!response.isSuccessful) {
+            throw RequestFailedException(
+                response.body,
```

Comment:
This is a message in RequestFailedException. Does response.body contain an error message? I think we could just add payload field to the RequestFailedException to avoid confusion.

Reply:
We could include HTTP status message in the response so we get errors like `RequestFailedException("Internal server error"` etc. 

According to the okhttp3 Response docs
```
/**
   * Returns a non-null value if this response was passed to [Callback.onResponse] or returned
   * from [Call.execute]. Response bodies must be [closed][ResponseBody] and may
   * be consumed only once.
   *
   * This always returns null on responses returned from [cacheResponse], [networkResponse],
   * and [priorResponse].
   */
```
body should always be present (we are checking it in the service to be sure), but I guess it can be an empty string sometimes.

Reply:
I just mean that the body might not be the message for the reason it all failed. I would rather add "payload" field there.